File: Tp03/TP03Q01 - Lista com Alocação Sequencial em Java/Principal.java
import java.util.*;


    public class Principal{
	   protected static Jogador[] jogadores;
	   protected static Lista lista;


        /**
         *@param vetor vazio

         metodo le o arquivo e preenche o vetor
         * */
        public static void leArquivoInicial(Jogador[] jogadores){

            Arq.openRead("/tmp/players.csv");
            String dados = "";
            int i = 0;

            Arq.readLine();//remove linha de cabecalho
            


            while(i<3922){
                dados = Arq.readLine();
                jogadores[i] = new Jogador();
                jogadores[i].ler(dados);

                i++;
            }
        }
        
        
        
        public static void SelecionaNovoVetor(int id, Jogador Novo){
        
        	Novo = jogadores[id].clone();
        
        
        }
        public static boolean isFim(String palavra){

            return palavra.length() == 3 && palavra.charAt(0)=='F' && palavra.charAt(1)=='I' && palavra.charAt(2)=='M';
        }
        
        
        
   		
   		
   		public static long now(){


			return new Date().getTime();

		}
		
		public static void descobreOperacao(String dado)throws Exception{
			String[] divisao = dado.split(" ",3);
			int posicao = 0;
			if(divisao.length > 1)
				posicao = Integer.parseInt(divisao[1]);
			Jogador retorno;
			switch (divisao[0]){
				case "II":
					lista.inserirInicio(jogadores[posicao]);
					break;
				case "I*":
					lista.inserir(jogadores[Integer.parseInt(divisao[2])],posicao);
					break;
				case "IF":
					lista.inserirFim(jogadores[posicao]);
					break;
				case "RI":
					retorno = lista.removerInicio();
					MyIO.println("(R) " + retorno.getNome());
					break;
				case "R*":
					retorno = lista.remover(Integer.parseInt(divisao[1]));
					MyIO.println("(R) " + retorno.getNome());
					break;
				case "RF":
					retorno = lista.removerFim();
					MyIO.println("(R) " + retorno.getNome());
					break;
			
			
			}
		
		}

        


        public static void main(String[] Args)throws Exception{
            int tam = 3922;

            jogadores = new Jogador[tam];//vetor original
			lista = new Lista();
            



            leArquivoInicial(jogadores);
            
            int posicao = 0;
            
            String palavra = MyIO.readString();

            while(!isFim(palavra)){
                posicao = Integer.parseInt(palavra);
                
                lista.inserirFim(jogadores[posicao]);
                
                palavra = MyIO.readString();
                     
            }

            long inicio = now();
            int operacoes = MyIO.readInt();//quantidade de operacoes na Estrutura
            
            for(int i = 0;i<operacoes;i++){
            	String dado = MyIO.readLine();
            
            	descobreOperacao(dado);	
            }
            
          	lista.mostrar();

        }






    }
    
class Jogador{

    private int id,
                altura,
                peso,
                anoNascimento;
    private String nome,
                   universidade,
                   cidadeNascimento,
                   estadoNascimento;

    public Jogador(){

    }
    public Jogador(int id, String nome, int altura,int peso,String universidade,int anoNascimento,String cidadeNascimento,String estadoNascimento){

        setId(id);
        setNome(nome);
        setAltura(altura);
        setPeso(peso);
        setUniversidade(universidade);
        setAnoNascimento(anoNascimento);
        setCidadeNascimento(cidadeNascimento);
        setEstadoNascimento(estadoNascimento);
    }
    public int getId(){
        return id;

    }
    public void setId(int id){
        if(id >= 0)
            this.id = id;
    }
    public String getNome(){
        return nome;

    }
    public void setNome(String nome){
        nome = validaDado(nome);

        this.nome = nome;
    }
    public int getAltura(){

        return altura;

    }
    public void setAltura(int altura){

        this.altura = altura;
    }
    public int getPeso(){

        return peso;

    }
    public void setPeso(int peso){

        this.peso = peso;


    }
    public String getUniversidade(){
        return universidade;
        
    }
    public void setUniversidade(String universidade){
        universidade = validaDado(universidade);
        this.universidade = universidade;
    }
    public int getAnoNascimento(){
        return anoNascimento;

    }
    public void setAnoNascimento(int anoNascimento){
        this.anoNascimento = anoNascimento;

    }
    public String getCidadeNascimento(){
        return cidadeNascimento;
        
    }
    public void setCidadeNascimento(String cidadeNascimento){
        cidadeNascimento = validaDado(cidadeNascimento);
        this.cidadeNascimento = cidadeNascimento;    

    }
    public String getEstadoNascimento(){
        return estadoNascimento;
        
    }
    public void setEstadoNascimento(String estadoNascimento){
        estadoNascimento = validaDado(estadoNascimento);
    
        this.estadoNascimento = estadoNascimento;
    }
    public String validaDado(String dado){

        if(dado.isEmpty())
            dado = "nao informado";

        return dado;

    }
    /**
     *Imprime a instancia
     * */

    public void imprimir(){

        MyIO.println("## " + getNome() + " ## " +getAltura() + " ## " + getPeso() + " ## " + getAnoNascimento() + " ## " + getUniversidade() + " ## " + getCidadeNascimento() + " ## "+ getEstadoNascimento() +" ##");
    }

    public void ler(String jogador){

        String[] valores = jogador.split(",",8);

        setId(valores[0]!="" ?Integer.parseInt(valores[0]): -1);
        setNome(valores[1]);//.replaceAll("\\*",""));
        setAltura(Integer.parseInt(valores[2]));
        setPeso(Integer.parseInt(valores[3]));
        setUniversidade(valores[4]);
        setAnoNascimento(Integer.parseInt(valores[5]));
        setCidadeNascimento(valores[6]);
        setEstadoNascimento(valores[7]);

    }
    public Jogador clone() {
        Jogador novo = new Jogador();
        novo.id = this.id;
        novo.nome = this.nome;
        novo.altura = this.altura;
        novo.peso = this.peso;
        novo.anoNascimento = this.anoNascimento;
        novo.universidade = this.universidade;
        novo.cidadeNascimento = this.cidadeNascimento;
        novo.estadoNascimento = this.estadoNascimento;
        return novo;
    }
    
    

}
    
/**
 * Lista estatica
 * @author Max do Val Machado
 * @version 2 01/2015
 */
class Lista {
   private Jogador[] array;
   private int n;


   /**
    * Construtor da classe.
    */
   public Lista () {
      this(3922);
   }


   /**
    * Construtor da classe.
    * @param tamanho Tamanho da lista.
    */
   public Lista (int tamanho){
      array = new Jogador[tamanho];
      n = 0;
   }


   /**
    * Insere um elemento na primeira posicao da lista e move os demais
    * elementos para o fim da lista.
    * @param x int elemento a ser inserido.
    * @throws Exception Se a lista estiver cheia.
    */
   public void inserirInicio(Jogador x) throws Exception {

      //validar insercao
      if(n >= array.length){
         throw new Exception("Erro ao inserir!");
      } 

      //levar elementos para o fim do array
      for(int i = n; i > 0; i--){
         array[i] = array[i-1].clone();
      }

      array[0] = x.clone();
      n++;
   }


   /**
    * Insere um elemento na ultima posicao da lista.
    * @param x int elemento a ser inserido.
    * @throws Exception Se a lista estiver cheia.
    */
   public void inserirFim(Jogador x) throws Exception {

      //validar insercao
      if(n >= array.length){
         throw new Exception("Erro ao inserir!");
      }

      array[n] = x.clone();
      n++;
   }


   /**
    * Insere um elemento em uma posicao especifica e move os demais
    * elementos para o fim da lista.
    * @param x int elemento a ser inserido.
    * @param pos Posicao de insercao.
    * @throws Exception Se a lista estiver cheia ou a posicao invalida.
    */
   public void inserir(Jogador x, int pos) throws Exception {

      //validar insercao
      if(n >= array.length || pos < 0 || pos > n){
         throw new Exception("Erro ao inserir!");
      }

      //levar elementos para o fim do array
      for(int i = n; i > pos; i--){
         array[i] = array[i-1].clone();
      }

      array[pos] = x.clone();
      n++;
   }


   /**
    * Remove um elemento da primeira posicao da lista e movimenta 
    * os demais elementos para o inicio da mesma.
    * @return resp int elemento a ser removido.
    * @throws Exception Se a lista estiver vazia.
    */
   public Jogador removerInicio() throws Exception {

      //validar remocao
      if (n == 0) {
         throw new Exception("Erro ao remover!");
      }

      Jogador resp = array[0].clone();
      n--;

      for(int i = 0; i < n; i++){
         array[i] = array[i+1].clone();
      }

      return resp;
   }


   /**
    * Remove um elemento da ultima posicao da lista.
    * @return resp int elemento a ser removido.
    * @throws Exception Se a lista estiver vazia.
    */
   public Jogador removerFim() throws Exception {

      //validar remocao
      if (n == 0) {
         throw new Exception("Erro ao remover!");
      }

      return array[--n];
   }


   /**
    * Remove um elemento de uma posicao especifica da lista e 
    * movimenta os demais elementos para o inicio da mesma.
    * @param pos Posicao de remocao.
    * @return resp int elemento a ser removido.
    * @throws Exception Se a lista estiver vazia ou a posicao for invalida.
    */
   public Jogador remover(int pos) throws Exception {

      //validar remocao
      if (n == 0 || pos < 0 || pos >= n) {
         throw new Exception("Erro ao remover!");
      }

      Jogador resp = array[pos].clone();
      n--;

      for(int i = pos; i < n; i++){
         array[i] = array[i+1];
      }

      return resp;
   }


   /**
    * Mostra os elementos da lista separados por espacos.
    */
   public void mostrar (){

      for(int i = 0; i < n; i++){
         System.out.print("[" + i + "] ");
         array[i].imprimir();
      }

   }


   /**
    * Procura um elemento e retorna se ele existe.
    * @param x int elemento a ser pesquisado.
    * @return <code>true</code> se o array existir,
    * <code>false</code> em caso contrario.
    */
   public boolean pesquisar(int x) {
      boolean retorno = false;
      for (int i = 0; i < n && retorno == false; i++) {
         retorno = (array[i].getId() == x);
      }
      return retorno;
   }
}





















    

File: Tp03/TP03Q02 - Lista com Alocação Sequencial em C/principal.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

typedef struct
{

    int id,
        altura,
        peso,
        anoNascimento;

    char *nome,
        *universidade,
        *cidadeNascimento,
        *estadoNascimento;

} Jogador;

typedef Jogador *PJogador;
typedef Jogador NJogador;

//cabecalhos "classe" Jogador
void construtor1(PJogador jogador);
char *trataAsterisco(char *str);
void ler(PJogador jogador, char dados[]);
void imprimir(PJogador jogador);
void construtor2(PJogador jogador, char dados[]);
void leArquivoInicial(PJogador jogador[]);
bool isFim(char *str);

//cabecalhos "classe" lista
void start();
void inserirInicio(PJogador x);
void inserirFim(PJogador x);
void inserir(PJogador x, int pos);
NJogador removerInicio();
NJogador removerFim();
NJogador remover(int pos);
void mostrar();

//cabecalhos "classe main"
void descobreOperacao(char entrada[], PJogador jogadores[]);

//variaveis

int main()
{

    PJogador jogadores[3922]; //array principal

    clock_t inicio, fim; //calcular tempo

    int pos = 0;
    char entrada[100];

    leArquivoInicial(jogadores);
    scanf(" %[^\n]", entrada);

    while (!isFim(entrada))
    {
        pos = atoi(entrada);
        inserirFim(jogadores[pos]);
        scanf(" %[^\n]", entrada);
    }
    int operacoes = 0;
    scanf(" %i", &operacoes);
    for (int i = 0; i < operacoes; i++)
    {
        //fflush(stdin);
        scanf(" %[^\n]", entrada);
        descobreOperacao(entrada, jogadores);
    }
    mostrar();
    return 0;
}

void descobreOperacao(char entrada[], PJogador jogadores[])
{
    char *divisao[3];
    divisao[0] = NULL;
    divisao[1] = NULL;
    divisao[2] = NULL;
    divisao[0] = strsep(&entrada, " ");
    divisao[1] = strsep(&entrada, " ");
    divisao[2] = strsep(&entrada, " ");
    NJogador retorno;

    int posicao = 0;
    if (divisao[1] != NULL)
    {
        posicao = atoi(divisao[1]);
    }

    if (strcmp(divisao[0], "II") == 0)
    {
        inserirInicio(jogadores[posicao]);
    }
    else if (strcmp(divisao[0], "I*") == 0)
    {
        inserir(jogadores[atoi(divisao[2])], posicao);
    }
    else if (strcmp(divisao[0], "IF") == 0)
    {
        inserirFim(jogadores[posicao]);
    }
    else if (strcmp(divisao[0], "RI") == 0)
    {
        retorno = removerInicio();
        printf("(R) %s\n", retorno.nome);
    }
    else if (strcmp(divisao[0], "R*") == 0)
    {
        retorno = remover(atoi(divisao[1]));
        printf("(R) %s\n", retorno.nome);
    }
    else if (strcmp(divisao[0], "RF") == 0)
    {
        retorno = removerFim();
        printf("(R) %s\n", retorno.nome);
    }
    else
    {

        printf("Erro ao selecionar Operacao\n");
    }
}
/**
/"Classe" Jogador
**/

void construtor1(PJogador jogador)
{

    jogador->id = 0;
    jogador->altura = 0;
    jogador->peso = 0;
    jogador->anoNascimento = 0;

    jogador->nome = (char *)malloc(sizeof(char) * 100);
    jogador->universidade = (char *)malloc(sizeof(char) * 100);
    jogador->cidadeNascimento = (char *)malloc(sizeof(char) * 100);
    jogador->estadoNascimento = (char *)malloc(sizeof(char) * 100);
}

void construtor2(PJogador jogador, char dados[])
{
    jogador->id = 0;
    jogador->altura = 0;
    jogador->peso = 0;
    jogador->anoNascimento = 0;

    jogador->nome = (char *)malloc(sizeof(char) * 100);
    jogador->universidade = (char *)malloc(sizeof(char) * 100);
    jogador->cidadeNascimento = (char *)malloc(sizeof(char) * 100);
    jogador->estadoNascimento = (char *)malloc(sizeof(char) * 100);
    ler(jogador, dados);
}

/**
	retira possiveis chiados nas strings
**/
char *trataAsterisco(char *str)
{

    if (strstr(str, "*") != NULL)
    {

        for (int i = 0; i < strlen(str); i++)
        {
            if (str[i] == '*')
            {
                str[i] = '\0';
            }
        }
    }

    return str;
}
/**
	divide a entrada inteira em partes menores e as le para o jogador
**/

void ler(PJogador jogador, char dados[])
{

    char *array[8];

    char *emBranco = (char *)malloc(sizeof(char *) * 100);
    strcpy(emBranco, "nao informado");

    array[0] = strsep(&dados, ",");

    for (int i = 1; i < 8; i++)
    {
        array[i] = strsep(&dados, ",");
    }

    //array[1] = trataAsterisco(array[1]);

    jogador->id = atoi(array[0]);
    jogador->altura = atoi(array[2]);
    jogador->peso = atoi(array[3]);
    jogador->anoNascimento = atoi(array[5]);

    if (strcmp(array[1], "\0") == 0)
        jogador->nome = emBranco;
    else
        jogador->nome = array[1];

    if (strcmp(array[4], "\0") == 0)
        jogador->universidade = emBranco;
    else
        jogador->universidade = array[4];

    if (strcmp(array[6], "\0") == 0)
    {
        jogador->cidadeNascimento = emBranco;
    }
    else
    {
        jogador->cidadeNascimento = array[6];
    }

    if (strcmp(array[7], "\n") == 0)
        jogador->estadoNascimento = emBranco;
    else
        jogador->estadoNascimento = strsep(&array[7], "\n");
}

/**
	Imprime instancia de Jogador selecionado nos moldes escolhidos
**/

void imprimir(PJogador jogador)
{

    printf("## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n", jogador->nome, jogador->altura, jogador->peso, jogador->anoNascimento, jogador->universidade, jogador->cidadeNascimento, jogador->estadoNascimento);
}

/**
	Clona instancia passada para nova posicao
**/

NJogador clone(PJogador jogador)
{
    NJogador retorno;
    construtor1(&retorno);
    retorno.id = jogador->id;
    strcpy(retorno.nome, jogador->nome);
    retorno.altura = jogador->altura;
    retorno.peso = jogador->peso;
    retorno.anoNascimento = jogador->anoNascimento;
    strcpy(retorno.universidade, jogador->universidade);
    strcpy(retorno.cidadeNascimento, jogador->cidadeNascimento);
    strcpy(retorno.estadoNascimento, jogador->estadoNascimento);

    return retorno;
}

/**

	le todos os integrantes do arquivo e os instancia como jogador
**/

void leArquivoInicial(PJogador jogador[])
{
    FILE *Arq = fopen("/tmp/players.csv", "r");

    char *dados = (char *)malloc(sizeof(char) * 200);
    int i = 0;
    fgets(dados, 200, Arq);

    while (i < 3922)
    {
        dados = (char *)malloc(sizeof(char) * 200);
        fgets(dados, 200, Arq);

        jogador[i] = (PJogador)malloc(sizeof(NJogador));
        construtor2(jogador[i], dados);

        i++;
    }
    fclose(Arq);
}

bool isFim(char *str)
{

    return strlen(str) == 3 && str[0] == 'F' && str[1] == 'I' && str[2] == 'M';
}

/**
/"Classe Lista"
**/

#define MAXTAM 3922

NJogador array[MAXTAM]; // Elementos da pilha
int n;                  // Quantidade de array.

/**
 * Inicializacoes
 */
void start()
{
    n = 0;
}

/**
 * Insere um elemento na primeira posicao da lista e move os demais
 * elementos para o fim da 
 * @param x int elemento a ser inserido.
 */
void inserirInicio(PJogador x)
{
    int i;

    //validar insercao
    if (n >= MAXTAM)
    {
        printf("Erro ao inserir!");
        exit(1);
    }

    //levar elementos para o fim do array
    for (i = n; i > 0; i--)
    {
        array[i] = clone(&array[i - 1]);
    }

    array[0] = clone(x);
    n++;
}

/**
 * Insere um elemento na ultima posicao da 
 * @param x int elemento a ser inserido.
 */
void inserirFim(PJogador x)
{

    //validar insercao
    if (n >= MAXTAM)
    {
        printf("Erro ao inserir!");
        exit(1);
    }

    array[n] = clone(x);
    n++;
}

/**
 * Insere um elemento em uma posicao especifica e move os demais
 * elementos para o fim da 
 * @param x int elemento a ser inserido.
 * @param pos Posicao de insercao.
 */
void inserir(PJogador x, int pos)
{
    int i;

    //validar insercao
    if (n >= MAXTAM || pos < 0 || pos > n)
    {
        printf("Erro ao inserir!");
        exit(1);
    }

    //levar elementos para o fim do array
    for (i = n; i > pos; i--)
    {
        array[i] = clone(&array[i - 1]);
    }

    array[pos] = clone(x);
    n++;
}

/**
 * Remove um elemento da primeira posicao da lista e movimenta 
 * os demais elementos para o inicio da mesma.
 * @return resp int elemento a ser removido.
 */
NJogador removerInicio()
{
    int i;
    NJogador resp;

    //validar remocao
    if (n == 0)
    {
        printf("Erro ao remover!");
        exit(1);
    }

    resp = clone(&array[0]);
    n--;

    for (i = 0; i < n; i++)
    {
        array[i] = clone(&array[i + 1]);
    }

    return resp;
}

/**
 * Remove um elemento da ultima posicao da 
 * @return resp int elemento a ser removido.
 */
NJogador removerFim()
{

    //validar remocao
    if (n == 0)
    {
        printf("Erro ao remover!");
        exit(1);
    }

    return array[--n];
}

/**
 * Remove um elemento de uma posicao especifica da lista e 
 * movimenta os demais elementos para o inicio da mesma.
 * @param pos Posicao de remocao.
 * @return resp int elemento a ser removido.
 */
NJogador remover(int pos)
{
    int i;
    NJogador resp;

    //validar remocao
    if (n == 0 || pos < 0 || pos >= n)
    {
        printf("Erro ao remover!");
        exit(1);
    }

    resp = clone(&array[pos]);
    n--;

    for (i = pos; i < n; i++)
    {
        array[i] = clone(&array[i + 1]);
    }

    return resp;
}

/**
 * Mostra os array separados por espacos.
 */
void mostrar()
{
    int i;



    for (i = 0; i < n; i++)
    {
        printf("[%d] ", i);
        imprimir(&array[i]);
    }
}

/**
 * Procura um array e retorna se ele existe.
 * @param x int elemento a ser pesquisado.
 * @return <code>true</code> se o array existir,
 * <code>false</code> em caso contrario.
 */
bool pesquisar(PJogador x)
{
    bool retorno = false;
    int i;

    for (i = 0; i < n && retorno == false; i++)
    {
        retorno = (array[i].id == x->id);
    }
    return retorno;
}
